precisely/mapping_matchers.py

```diff
@@ -36,3 +38,21 @@ def describe(self):
             "{0!r}: {1}".format(key, matcher.describe())
             for key, matcher in self._matchers.items()
         )))
+
+
+def mapping_includes(matchers):
+    return MappingIncludesMatcher(dict(
+        (key, to_matcher(matcher))
+        for key, matcher in matchers.items()
+    ))
+
+
+class MappingIncludesMatcher(IsMappingMatcher):
+    def __init__(self, matchers):
+        super(MappingIncludesMatcher, self).__init__(matchers, allow_extra=True)
+
+    def describe(self):
+        return "mapping including items:{0}".format(indented_list(sorted(
+            "{0!r}: {1}".format(key, matcher.describe())
+            for key, matcher in self._matchers.items()
+        )))
```

Comment:
The split in logic here feels awkward to me: half the specialisation for `mapping_includes()` is in this subclass, while the other half is in the superclass. I think it'd be simpler to switch on `self._allow_extra` in `IsMappingMatcher.describe()`. Other than that, I think this looks good.

Reply:
I went with a slightly more subclassing-based approach and got rid of _allow_extra entirely. What do you think of the latest push?

Reply:
I tend to avoid implementation inheritance (or more specifically the template pattern). I'd suggest either switching on `self._allow_extra`, or creating two separate classes that both delegate to a common matching function.

Reply:
I've put `self._allow_extra` back in.